File: go/oop-patterns/behavioral-patterns/chain-of-responsibility/chain-of-responsibility.go
package main

import "fmt"

// интерфейс для определения методов назначения следующего обработчика
// и непосредственно обработки запроса
type handler interface {
	setNext(handler handler) handler
	handle(request string) string
}

// базовый тип, который содержит логику назначения следующего обработчика и
// базовую логику обработки
type abstractHandler struct {
	nextHandler handler
}

func (h *abstractHandler) setNext(handler handler) handler {
	h.nextHandler = handler
	return handler
}

func (h *abstractHandler) handle(request string) string {
	if h.nextHandler != nil {
		return h.nextHandler.handle(request)
	}
	return ""
}

// Конкретные обработчики либо обрабатывают запрос сами, либо передают его дальше
// по цепочке обработчиков
type monkeyHandler struct {
	abstractHandler
}

func (h *monkeyHandler) handle(request string) string {
	if request == "Банан" {
		return "Обезьяна: Я съем Банан."
	}
	return h.abstractHandler.handle(request)
}

type squirrelHandler struct {
	abstractHandler
}

func (h *squirrelHandler) handle(request string) string {
	if request == "Орех" {
		return "Белка: Я съем Орех."
	}
	return h.abstractHandler.handle(request)
}

type dogHandler struct {
	abstractHandler
}

func (h *dogHandler) handle(request string) string {
	if request == "Мясо" {
		return "Собака: Я съем Мясо."
	}

	return h.abstractHandler.handle(request)
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	monkey := monkeyHandler{}
	squirrel := squirrelHandler{}
	dog := dogHandler{}

	monkey.setNext(&squirrel).setNext(&dog)

	handler := monkey

	foods := []string{"Орех", "Банан", "Кофе"}

	for _, food := range foods {
		fmt.Printf("Клиент: кто хочет %v?\n", food)

		result := handler.handle(food)
		if result != "" {
			fmt.Println(result)
		} else {
			fmt.Printf("%v никто не захотел.\n", food)
		}
	}
}

File: javascript/oop-patterns/behavioral-patterns/chain-of-responsibility/chain-of-responsibility.js
// базовый класс, который содержит логику назначения следующего обработчика и
// базовую логику обработки
class AbstractHandler {
  nextHandler;

  setNext(handler) {
    this.nextHandler = handler;
    return handler;
  }

  handle(request) {
    if (this.nextHandler) {
      return this.nextHandler.handle(request);
    }
    return null;
  }
}

// Конкретные обработчики либо обрабатывают запрос сами, либо передают его дальше
// по цепочке обработчиков
class MonkeyHandler extends AbstractHandler {
  handle(request) {
    if (request === 'Банан') {
      return `Обезьяна: Я съем ${request}.`;
    }
    return super.handle(request);
  }
}

class SquirrelHandler extends AbstractHandler {
  handle(request) {
    if (request === 'Орех') {
      return `Белка: Я съем ${request}.`;
    }
    return super.handle(request);
  }
}

class DogHandler extends AbstractHandler {
  handle(request) {
    if (request === 'Мясо') {
      return `Собака: Я съем ${request}.`;
    }
    return super.handle(request);
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
const monkey = new MonkeyHandler();
const squirrel = new SquirrelHandler();
const dog = new DogHandler();

monkey.setNext(squirrel).setNext(dog);

const handler = monkey;

const foods = ['Орех', 'Банан', 'Кофе'];

for (const food of foods) {
  console.log(`Клиент: Кто хочет ${food}?`);
  const result = handler.handle(food);
  if (result) {
    console.log(`  ${result}`);
  } else {
    console.log(`  ${food} никто не захотел.`);
  }
}
